docsite/rst/developing_program_flow_modules.rst

```diff
@@ -0,0 +1,392 @@
+* Handles any special cases regarding modules (for instance, various
+  complications around Windows modules that need to have the same names as
+  python modules so that internal calling of modules from other Action Plugins
+  work.)
+
+Much of this functionality comes from the :code:`BaseAction`BaseAction class
+which lives in :file:`plugins/action/__init__.py`.  It makes use of Connection
+and Shell objects to do its work.
+
+.. _flow_executor_module_common:
+
+executor/module_common.py
+-------------------------
+
+Code in :file:`executor/module_common.py` takes care of assembling the module
+to be shipped to the managed node.  The module is first read in, then examined
+to determine its type.  :ref:`PowerShell <flow_powershell_modules>` and
+:ref:`JSON-args modules <flow_jsonargs_modules>` are passed through
+:ref:`Module Replacer <module_replacer>`.  New-style
+:ref:`python modules <flow_python_modules>` are are assembled by
+:ref:`ziploader`.
+:ref:`Non-native-want-json <flow_want_json_modules>` and
+:ref:`Old-Style modules <flow_old_style_modules>` aren't touched by either of
+these and pass through unchanged.  After the assembling step, one final
+modification is made to all modules that have a shebang line.  We check
+whether the interpreter in the shebang line has a specific path configured via
+an ``ansible_$X_interpreter`` inventory variable.  If it does we substitute that
+path for the interpreter path given in the module.  After this we return the
+complete module data and the module type to the Normal Action which continues
+execution of the module.
+
+Next we'll go into some details of the two assembler frameworks.
+
+.. _module_replacer:
+
+module replacer
+^^^^^^^^^^^^^^^
+
+Module replacer is essentially a preprocessor (like the C Preprocessor for
+those familiar with that language).  It does straight substitutions of specific
+substring patterns in the module file.  There are two types of substitutions:
+
+* Replacements that only happen in the module file.  These are public
+  replacement strings that modules can utilize to get helpful boilerplate or
+  access to arguments.
+
+  - :code:`from ansible.module_utils.MOD_LIB_NAME import *` is replaced with the
+    contents of the :file:`ansible/module_utils/MOD_LIB_NAME.py`  These should
+    only be used with :ref:`new-style python modules <flow_python_modules>`.
+  - :code:`#<<INCLUDE_ANSIBLE_MODULE_COMMON>>` is equivalent to
+    :code:`from ansible.module_utils.basic import *` and should also only apply
+    to new-style python modules.
+  - :code:`# POWERSHELL_COMMON` substitutes the contents of
+    :file:`ansible/module_utils/powershell.ps1`.  It should only be used with
+    :ref:`new-style Powershell modules <flow_powershell_modules>`.
+
+* Replacements that are used by ansible ``module_utils`` code.  These are internal
+  replacement patterns.  They may be used internally in the above public
+  replacements but shouldn't be used directly by modules.
```

Comment:
add two commas:  They may be used internally, in the above public replacements, but shouldn't be used directly by modules.